### PR TITLE
fix pk may contains default null columns

### PR DESCRIPTION
**Pull request by xiongjiwei**
all part of the primary key must be `NOT NULL` and should contain all partition columns, so that if the table has a primary key, the partition column cannot be default null.

### CODE REVIEW

File: sqlgen/start.go [hunk 1]
```diff
@@ -193,22 +193,26 @@ func NewGenerator(state *State) func() string {
...
 						c := partitionedCol.ToColumn()
 						Assert(c != nil)
+						if idx.Tp == IndexTypePrimary {
+							if c.defaultVal == "null" {
+								return Empty()
+							}
+							idx.RemoveColIf(func(col *Column) bool {
+								return col.defaultVal == "null"
+							})
+						}
```
Comment (tangenta) [inline sqlgen/start.go]: Is it possible that all the primary key columns are removed?

Can we solve this in the `GenNewIndex()`?

Comment (xiongjiwei) [reply sqlgen/start.go]: I rewrite the code. PTAL again.

File: sqlgen/db_generator.go [hunk 2]
```diff
@@ -84,6 +84,9 @@ func GenNewIndex(id int, tbl *Table, w *Weight) *Index {
 		}
 		totalCols[0], totalCols[chosenIdx] = totalCols[chosenIdx], totalCols[0]
 		totalCols = totalCols[1:]
+		if idx.Tp == IndexTypePrimary && chosenCol.defaultVal == "null" {
+			continue
+		}
```
Comment (tangenta) [inline sqlgen/db_generator.go:89]: It is possible that all the columns are nullable, and the `GenNewIndex()` will panic.

File: sqlgen/start.go [hunk 3]
```diff
@@ -173,6 +173,9 @@ func NewGenerator(state *State) func() string {
 		colDefs = NewFn("colDefs", func() Fn {
 			colDef = NewFn("colDef", func() Fn {
 				col := GenNewColumn(state.AllocGlobalID(ScopeKeyColumnUniqID), w)
+				if len(tbl.Columns) == 0 && col.defaultVal == "null" {
+					col.defaultVal = ""
+				}
```
Comment (xiongjiwei) [inline sqlgen/start.go:178]: to quick fix this issue, there must have a not nullable column

File: sqlgen/db_generator.go (hunk 2 above)
Comment (xiongjiwei) [reply sqlgen/db_generator.go:89]: right, then the table cannot have partitions and cannot have a primary key. I have changed the `colDef` to make sure the table has at least one not nullable column.

**Comment by tangenta** [comment]
I think it is better to check the column's default values in `GenNewIndex()` instead of keeping the assumption "must have a not nullable column".

Fixed by https://github.com/PingCAP-QE/clustered-index-rand-test/commit/18f45c0d0c55ba3eaa894e89c55884982ee9ccaf.

**Pull request closed without merging**